RFCs/FS-1142-Extended-numeric-literal.md

```diff
@@ -0,0 +1,137 @@
+    let x13 = 0o_52            // OK (Invalid *previously*)
+    let x14 = 0o5_2            // OK (octal literal)
+    let x15 = 0o52_            // Invalid cannot put underscores at the end of a number
+    ```
+
+2. Allow number prefix (`0x`, `0o`, `0b`) before integer custom numeric literals.
+  
+    ```fsharp
+    let x1 = 0x123I        // big int (291)
+    let x2 = 0o123I        // big int (83)
+    let x3 = 0b1010I       // big int (10)
+    ```
+
+    The number prefix **will not be removed** from the string sended to the `FromString` of the numeric literal module. This may be a break change.
+
+    ```fsharp
+    module NumericLiteralG =
+      let FromString s = s
+
+    let x4 = 0x_99_9999_9999_9999_9999G    // x4 will be "0x999999999999999999"
+    ```
+
+3. Allow floating point custom numeric literals. 
+
+    This will require the numeric literal module contains two new functions:
+
+    ```fsharp
+    val FromFloat: float -> 'CustomNumber
+    val FromFloatString: string -> 'CustomNumber
+    ```
+
+    According to [this comment](https://github.com/fsharp/fslang-suggestions/issues/445#issuecomment-596902041),
+
+    - When the literal has <= 15 significant figures[^1] and its exponent is within -300 to 300, the compiler will call `FromFloat` with the parsed `float` number
+    - Or the compiler will call `FromFloatString` with the original string
+
+    [^1]: The first none zero figure to last none zero figure, without the dot. Can be match by the Regex:
+
+        ```regex
+        ^-?0*(?<number>\d+\.?\d*?)0*(?:$|[eE][+-]?(?<exp>\d+))
+        ```
+
+    ```fsharp
+    module NumericLiteralG =
+      let FromFloat (s: float) = s
+      let FromFloatString (s: string) = s
+
+    let x1 = 123.456e-10G    // x1: float = 1.23456e-08
+    let x2 = 123.456789123456789123e-10G    // x2: string = "123.456789123456789123e-10"
+    ```
+
+# Drawbacks
+
+- Custom numeric literals module might harder to write.
+
+- May introduce break change.
+
+# Alternatives
+
+- For number prefix (`0x`, `0o`, `0b`) before integer custom numeric literals, we might introduce a new `FromIntegerString` to avoid the break change.
+- Or firstly parse it to `bigint` then `ToString` to obtain a literal without prefix.
+
+# Compatibility
+
+Please address all necessary compatibility questions:
+
+* Is this a breaking change?
+Maybe
+
+* What happens when previous versions of the F# compiler encounter this design addition as source code?
+Can write numeric literal module with new functions in the source code, but cannot use these new numeric literal grammar.
+
+* What happens when previous versions of the F# compiler encounter this design addition in compiled binaries?
+Cannot use the new numeric literal grammar.
+
+# Pragmatics
+
+## Tooling
+
+Please list the reasonable expectations for tooling for this feature, including any of these:
+
+* Colorization
+
+  Might need to change the color schema of the numeric literals.
```

Comment:
this is probably not going to be a problem, as the syntax highlighter uses the parser results. If the parser says something is a numeric literal, it will be colored as such.

But certainly something to test before merging this change